src/librustc/traits/error_reporting.rs

```diff
@@ -858,26 +884,42 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
     }
 
 
-    fn need_type_info(&self, span: Span, ty: Ty<'tcx>) {
+    fn need_type_info(&self,
+                      obligation: &PredicateObligation<'tcx>,
+                      ty: Ty<'tcx>) {
+
         let ty = self.resolve_type_vars_if_possible(&ty);
-        let name = if let ty::TyInfer(ty::TyVar(ty_vid)) = ty.sty {
-            let ty_vars = self.type_variables.borrow();
-            if let TypeVariableOrigin::TypeParameterDefinition(_, name) =
-                    *ty_vars.var_origin(ty_vid)
-            {
-                name.to_string()
+        let ref cause = obligation.cause;
+
+        let mut err = struct_span_err!(self.tcx.sess,
+                                       cause.span,
+                                       E0282,
+                                       "unable to fully infer type(s)");
```

Comment:
Some possible things we could say:

- "unable to fully infer types" -- I don't think the (s) is needed. There are always lots of types at play, a plural is always appropriate. =)
- "type annotations needed"
- "multiple possible types detected; annotations needed to clarify" -- or something like that?
    - I'd like to get across that rustc is refusing to guess what your program means, basically, by picking a random type on your behalf.